sdc/rabbit/consumers.py

```diff
@@ -383,40 +383,74 @@ def stop(self):
         logger.info('Stopped')
 
+    def _delay_before_reconnect(self):
+        """Sleeps for a number of seconds equal to the `count` attribute on this object
+        """
+        logger.info("Sleeping before reconnect", no_of_seconds=self._count)
+        time.sleep(self._count)
+        self._count += 1
+        logger.info("Finished sleeping")
+
 
 class TornadoConsumer(AsyncConsumer):
+    """This is a consumer that uses the AsyncConsumer as a base but uses TornadoConnection
+    to connect to RabbitMQ
+    """
     def connect(self):
         """This method connects to RabbitMQ using a TornadoConnection object,
         returning the connection handle.
 
         When the connection is established, the on_connection_open method
         will be invoked by pika.
 
-        :rtype: pika.SelectConnection
+        :rtype: pika.adapters.TornadoConnection
 
         """
-        count = 1
         no_of_servers = len(self._rabbit_urls)
 
         while True:
-            server_choice = (count % no_of_servers) - 1
+            server_choice = (self._count % no_of_servers) - 1
 
             self._url = self._rabbit_urls[server_choice]
 
             try:
-                logger.info('Connecting', attempt=count)
+                logger.info('Connecting', attempt=self._count)
                 return TornadoConnection(pika.URLParameters(self._url),
                                          on_open_callback=self.on_connection_open,
                                          on_open_error_callback=self.on_connection_open_error,
                                          on_close_callback=self.on_connection_closed)
             except pika.exceptions.AMQPConnectionError:
                 logger.exception("Connection error")
-                count += 1
-                logger.error("Connection sleep", no_of_seconds=count)
-                time.sleep(count)
+                self._delay_before_reconnect()
 
                 continue
 
+    def close_connection(self):
+        """This method closes the connection to RabbitMQ."""
+        if self._connection.is_closing or self._connection.is_closed:
+            logger.info('Connection is closing or already closed', state=self._connection.connection_state)
+        else:
+            logger.info('Closing connection', state=self._connection.connection_state)
+            self._connection.close()
+
+    def on_connection_closed(self, _unused_connection, reason):
+        """This method is invoked by pika when the connection to RabbitMQ is
+        closed unexpectedly. Since it is unexpected, we will reconnect to
+        RabbitMQ if it disconnects.
+
+        :param pika.connection.Connection _unused_connection: The closed connection obj
+        :param Exception reason: exception representing reason for loss of
+            connection.
+
+        """
+        self._channel = None
+        if self._closing:
+            logger.warning("Connection closed, stopping ioloop", reason=reason)
```

Comment:
Is this best at `WARN`? A situation where we're in the process of deliberately closing the connection but warn that we've unexpectedly closed it is an odd one!

Reply:
I copied it from https://pika.readthedocs.io/en/stable/examples/tornado_consumer.html 
I think warning seems about right as it could indicate something going wrong, but if it reconnects straight away without issue then there was no need to say anything was wrong.  Does that seem reasonable to you?

Reply:
Worth noting too that the `logger` levels are _necessarily_ directly mapped to the POSIX levels.